jax/interpreters/xla.py

```diff
@@ -45,37 +45,100 @@
                   strtobool(os.getenv('JAX_DEBUG_NANS', "False")),
                   'Add nan checks to every operation.')
 
+def _map(f, *xs): return tuple(map(f, *xs))
+def identity(x): return x
+
+
+### handlers
+
+xb.register_constant_handler(core.Unit, lambda c, *_: c.Tuple())
+
+def aval_to_xla_shape(aval):
+  try:
+    return xla_shape_handlers[type(aval)](aval)
+  except KeyError:
+    raise TypeError("No xla_shape_handler for type: {}".format(type(aval)))
+xla_shape_handlers = {}
+xla_shape_handlers[core.AbstractUnit] = lambda _: xc.Shape.tuple_shape(())
+xla_shape_handlers[ShapedArray] = lambda a: xc.Shape.array_shape(a.dtype, a.shape)
+xla_shape_handlers[ConcreteArray] = lambda a: xc.Shape.array_shape(a.dtype, a.shape)
+
+def aval_to_result_handler(aval):
+  try:
+    return xla_result_handlers[type(aval)](aval)
+  except KeyError:
+    raise TypeError("No xla_result_handler for type: {}".format(type(aval)))
+xla_result_handlers = {}
+xla_result_handlers[core.AbstractUnit] = lambda _: lambda _: core.unit
+def array_result_handler(aval): return partial(DeviceArray, aval)
+xla_result_handlers[ShapedArray] = array_result_handler
+xla_result_handlers[ConcreteArray] = array_result_handler
+
+def device_put(x, device_num=0):
+  x = canonicalize_dtype(x)
+  try:
+    return device_put_handlers[type(x)](x, device_num)
+  except KeyError:
+    raise TypeError("No device_put handler for type: {}".format(type(x)))
+device_put_handlers = {}
+device_put_handlers[core.Unit] = \
+    lambda _, n: xc.Buffer.from_pyval((), n, backend=xb.get_backend())
```

Comment:
Why does unit need a device representation?

Reply:
Good question; we could try deleting this! But I think it's needed because it's a JAX value that can be produced and consumed by computations, especially those coming from partial evaluation. We use them as placeholders/sentinels when doing partial eval, so that for example if we do a partial-eval of a function like

```python
def f(x, y):
  return (x, y)
```

where we know `x=2` but don't yet know `y`, then the output of the first stage is represented as `[2, unit]` and the output of the second stage is represented as the jaxpr equivalent of `lambda x, y: [unit, y]` (to be applied to inputs like `[unit, 3]`). Then `merge_pvals` is what can join the two outputs.

Since those computations could be `jit`-compiled, we need to be able to handle units.

Do you buy that?

Reply:
Yeah, I see how taking out the units when lowering to jaxpr/xla would require a lot more bookkeeping about argument indices in things like partial_eval.